File: Binary-Search/0162-find-peak-element.py
# Time  : O(logn)
# Space : O(1)

class Solution:
  def findPeakElement(self, A: List[int]) -> int:
    n = len(A)

    if n == 1:
      return 0
    if A[0] > A[1]:
      return 0
    if A[n-2] < A[n-1]:
      return n-1

    lo, hi = 1, n - 2

    while lo <= hi:
      mid = lo + (hi - lo) // 2
      if A[mid - 1] < A[mid] > A[mid + 1]:  
        return mid
      elif A[mid] < A[mid + 1]:
        lo = mid + 1
      else:
        hi = mid - 1

# Time  : O(logn)
# Space : O(1)

class Solution:
  def findPeakElement(self, A: List[int]) -> int:
    lo, hi = 0, len(A) - 1

    while lo < hi:
      mid = lo + (hi - lo) // 2
      if A[mid] < A[mid + 1]:
        lo = mid + 1
      else:
        hi = mid

    return lo

File: Binary-Search/0162-find-peak-element.cpp
// Time  : O(logn)
// Space : O(1)
class Solution {
public:
  int findPeakElement(vector<int>& A) {
    int lo = 0;
    int hi = A.size()-1;
    while (lo < hi) {
      int mid = (lo + hi)>>1;
      if (A[mid] < A[mid+1]) lo = mid+1;
      else hi = mid;
    }
    return lo;
  }
};